src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -1,32 +1,165 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE DeriveFunctor       #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
 module Database.LSMTree.Internal.Lookup (
-    Run
+    -- * Lookup preparation
+    RunLookupView (..)
+  , RunIx
+  , KeyIx
   , prepLookups
+  , bloomQueries
+  , indexSearches
   ) where
 
-import           Data.Foldable (Foldable (..))
+import           Control.Monad.Primitive
+import           Control.Monad.ST.Strict
+import           Data.Primitive.ByteArray
+import           Data.Primitive.PrimVar
+import qualified Data.Vector as V
+import qualified Data.Vector.Mutable as VM
+import qualified Data.Vector.Unboxed as VU
+import qualified Data.Vector.Unboxed.Mutable as VUM
 import           Database.LSMTree.Internal.Run.BloomFilter (Bloom)
 import qualified Database.LSMTree.Internal.Run.BloomFilter as Bloom
 import           Database.LSMTree.Internal.Run.Index.Compact (CompactIndex,
-                     PageSpan)
+                     PageSpan (..))
 import qualified Database.LSMTree.Internal.Run.Index.Compact as Index
 import           Database.LSMTree.Internal.Serialise
+import           System.FS.API (Handle)
+import           System.FS.BlockIO.API
+
+-- | View of a 'Run' with only the essentials for lookups.
+data RunLookupView h = RunLookupView {
+    rlvKOpsHandle :: !h
+  , rlvBloom      :: !(Bloom SerialisedKey)
+  , rlvIndex      :: !CompactIndex
+  }
+  deriving Functor
 
--- | TODO: placeholder type for a run, replace by actual type once implemented
-type Run fd = (fd, Bloom SerialisedKey, CompactIndex)
+-- | Prepare disk lookups by doing bloom filter queries, index searches and
+-- creating 'IOOp's. The result is a vector of 'IOOp's and a vector of indexes,
+-- both of which are the same length. The indexes record the run and key
+-- associated with each 'IOOp'.
+prepLookups ::
+     PrimMonad m
```

Comment:
I've specialised for both, because I'm not sure yet which version we would want. In #102, I'm calling these functions in `IO`, and I did add a top-level `SPECIALIZE` pragma there, but I forgot to add them here

Reply:
Ok.